File: packages/sentinel-js/src/schema.ts
export function validateSchema(
  schema: Record<string, unknown>,
  value: unknown,
  path = 'root'
): string[] {
  const errors: string[] = []
  const expectedType = schema.type as string | undefined

  if (expectedType !== undefined && !matchesType(value, expectedType)) {
    errors.push(`${path}: expected type '${expectedType}', got '${typeName(value)}'`)
    return errors
  }

  if (Array.isArray(schema.enum) && !schema.enum.includes(value)) {
    errors.push(
      `${path}: value ${JSON.stringify(value)} not in enum ${JSON.stringify(schema.enum)}`
    )
  }

  if (expectedType === 'object' && isPlainObject(value)) {
    const required = Array.isArray(schema.required) ? (schema.required as string[]) : []
    for (const key of required) {
      if (!(key in value)) errors.push(`${path}.${key}: required property missing`)
    }

    const properties = isPlainObject(schema.properties)
      ? (schema.properties as Record<string, unknown>)
      : {}
    for (const [key, subSchema] of Object.entries(properties)) {
      if (key in value) {
        errors.push(
          ...validateSchema(subSchema as Record<string, unknown>, value[key], `${path}.${key}`)
        )
      }
    }
  }

  if (expectedType === 'array' && Array.isArray(value) && schema.items) {
    value.forEach((item, index) => {
      errors.push(
        ...validateSchema(schema.items as Record<string, unknown>, item, `${path}[${index}]`)
      )
    })
  }

  return errors
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function matchesType(value: unknown, expectedType: string): boolean {
  switch (expectedType) {
    case 'string':
      return typeof value === 'string'
    case 'number':
      return typeof value === 'number'
    case 'integer':
      return typeof value === 'number' && Number.isInteger(value)
    case 'boolean':
      return typeof value === 'boolean'
    case 'object':
      return isPlainObject(value)
    case 'array':
      return Array.isArray(value)
    case 'null':
      return value === null
    default:
      return true
  }
}

function typeName(value: unknown): string {
  if (value === null) return 'null'
  if (Array.isArray(value)) return 'array'
  if (typeof value === 'number') return Number.isInteger(value) ? 'integer' : 'number'
  return typeof value
}

File: packages/sentinel-py/sentinel/_schema.py
from typing import Any, List


def validate_schema(schema: dict, value: Any, path: str = "root") -> List[str]:
    """Validate `value` against a minimal JSON Schema subset.

    Supports: type, required, properties, items, enum. Returns a list of
    human-readable error strings; an empty list means the value is valid.
    """
    errors: List[str] = []

    expected_type = schema.get("type")
    if expected_type is not None and not _matches_type(value, expected_type):
        errors.append(f"{path}: expected type '{expected_type}', got '{_type_name(value)}'")
        return errors

    if "enum" in schema and value not in schema["enum"]:
        errors.append(f"{path}: value {value!r} not in enum {schema['enum']!r}")

    if expected_type == "object" and isinstance(value, dict):
        for required_key in schema.get("required", []):
            if required_key not in value:
                errors.append(f"{path}.{required_key}: required property missing")

        for key, sub_schema in schema.get("properties", {}).items():
            if key in value:
                errors.extend(validate_schema(sub_schema, value[key], f"{path}.{key}"))

    if expected_type == "array" and isinstance(value, list) and "items" in schema:
        for index, item in enumerate(value):
            errors.extend(validate_schema(schema["items"], item, f"{path}[{index}]"))

    return errors


_TYPE_CHECKS = {
    "string": lambda v: isinstance(v, str),
    "number": lambda v: isinstance(v, (int, float)) and not isinstance(v, bool),
    "integer": lambda v: isinstance(v, int) and not isinstance(v, bool),
    "boolean": lambda v: isinstance(v, bool),
    "object": lambda v: isinstance(v, dict),
    "array": lambda v: isinstance(v, list),
    "null": lambda v: v is None,
}


def _matches_type(value: Any, expected_type: str) -> bool:
    check = _TYPE_CHECKS.get(expected_type)
    return check(value) if check else True


def _type_name(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, int):
        return "integer"
    if isinstance(value, float):
        return "number"
    if isinstance(value, str):
        return "string"
    if isinstance(value, list):
        return "array"
    if isinstance(value, dict):
        return "object"
    return type(value).__name__
